Kernel/Syscalls/execve.cpp

```diff
@@ -851,15 +887,20 @@ ErrorOr<void> Process::exec(NonnullOwnPtr<KString> path, Vector<NonnullOwnPtr<KS
 
     VERIFY(description->inode());
 
-    // Read the first page of the program into memory so we can validate the binfmt of it
-    char first_page[PAGE_SIZE];
-    auto first_page_buffer = UserOrKernelBuffer::for_kernel_buffer((u8*)&first_page);
-    auto nread = TRY(description->read(first_page_buffer, sizeof(first_page)));
+    // Read just the size of ELF header of the program into memory so we can start parsing it
+    // If it is #! interpreted file, the size of a ELF header should still suffice.
+    auto preliminary_buffer = Array<u8, sizeof(Elf_Ehdr)>::from_repeated_value(0);
+    auto preliminary_read_buffer = UserOrKernelBuffer::for_kernel_buffer(preliminary_buffer.data());
+    auto nread = TRY(description->read(preliminary_read_buffer, 0, preliminary_buffer.span().size()));
 
     // 1) #! interpreted file
-    auto shebang_result = find_shebang_interpreter_for_executable(first_page, nread);
-    if (!shebang_result.is_error()) {
-        auto shebang_words = shebang_result.release_value();
+    if (is_executable_starting_with_shebang(preliminary_buffer)) {
+        // FIXME: PAGE_SIZE seems like enough for specifying an interpreter for now.
+        // We might need to re-evaluate this but to avoid further allocations, this is how it is for now.
+        auto shebang_line_buffer = Array<u8, PAGE_SIZE>::from_repeated_value(0);
+        auto shebang_line_read_buffer = UserOrKernelBuffer::for_kernel_buffer(shebang_line_buffer.data());
+        TRY(description->read(shebang_line_read_buffer, 0, shebang_line_buffer.span().size()));
+        auto shebang_words = TRY(find_shebang_interpreter_for_executable(shebang_line_buffer));
```

Comment:
We should probably error out in case we don't find `\n` in the buffer (i.e. `PAGE_SIZE` was not enough).

Reply:
Hmm, is there a standard in Unix (or Unix-like) systems to how all of this work actually? is `\n` actually the valid sign to determine this?

Reply:
I think I understand what you say here. It's interesting though to understand what happens if someone inserts something else (like `\0`). But probably this is good enough for now to just search for the new line break.

I sent a fix for this, please let me know if this is OK now.

Reply:
`\n` was just a placeholder for "whatever matches our interpretation of the shebang". In other words, what I was concerned about is the shebang being truncated accidentally (which it looks like this code does).

Reply:
Ah, OK. I added a check now for finding a new line break in the actual `find_shebang_interpreter_for_executable` function, so is it OK by now?